core/new-gui/src/app/workspace/service/workflow-collab/workflow-collab.service.ts

```diff
@@ -0,0 +1,67 @@
+import { Injectable } from "@angular/core";
+import { webSocket } from "rxjs/webSocket";
+import { CommandMessage } from "../workflow-graph/model/workflow-action.service";
+import { environment } from "../../../../environments/environment";
+import { Subject, Observable } from "rxjs";
+
+@Injectable({
+  providedIn: "root",
+})
+
+// This service will be for workflow collab
+// First step: move the arrays from undo/redo service into Yjs doc.
+// Once we have some stuff set I'll send the changes between clients and just see how to work with them.
+// Need to find a way to push as part of change, not update like that
+
+// How do we recover if the connection gets dropped?
+export class WorkflowCollabService {
+  private sendData: boolean = false; // set initial value to false to disable service
+
+  private socket = webSocket({
+    url: this.getWorkflowWebsocketUrl(),
+    deserializer: msg => msg["data"],
+  });
+
+  // maybe create a subject for testing?
+  private messageSubject: Subject<CommandMessage> = new Subject<CommandMessage>();
+
+  constructor() {
+    const self = this;
+    if (environment.enableWorkflowCollab) {
+      // to disable service, change to true to enable
+      this.setSendData(true);
+      this.socket.subscribe({
+        next(response) {
+          const message = JSON.parse(response) as CommandMessage;
+          self.messageSubject.next(message);
+        },
+        complete() {
+          console.log("websocket finished and disconnected");
+        },
+      });
+    }
+  }
+
+  public setSendData(toggle: boolean): void {
+    this.sendData = toggle;
+  }
+
+  public getSendData(): boolean {
+    return this.sendData;
+  }
+
+  public sendCommand(update: string): void {
+    this.socket.next(JSON.parse(update));
+  }
+
+  public getCommandMessageStream(): Observable<CommandMessage> {
+    return this.messageSubject.asObservable();
+  }
+
+  public getWorkflowWebsocketUrl(): string {
+    const websocketUrl = new URL("collaboration", document.baseURI);
+    // replace protocol, so that http -> ws, https -> wss
+    websocketUrl.protocol = websocketUrl.protocol.replace("http", "ws");
+    return websocketUrl.toString();
+  }
```

Comment:
this is a duplicate from the other websocket file. try to unify them and use it as a util function.

Reply:
These are the only two places to use this, don't need to create a util file for this.

Reply:
I still suggest you use one util function, for maintenance purpose. it is hard to remember all duplicated locations in case of updating them in the future.

Also, I thought the endpoint has changed to /wsapi/collab on the backend, here you are still hard coding it to collaboration, does it work?

Reply:
I've changed the endpoint, you're seeing the outdated file.